core/capabilities/remote/target_caller.go

```diff
@@ -0,0 +1,177 @@
+	}
+
+	deterministicMessageID := sha256.Sum256(rawRequest)
+
+	responseWaitGroup := &sync.WaitGroup{}
+	responseWaitGroup.Add(1)
+	c.messageIDToWaitgroup.Store(deterministicMessageID, responseWaitGroup)
+
+	responseReceived := make(chan struct{})
+	go func() {
+		responseWaitGroup.Wait()
+		close(responseReceived)
+	}()
+
+	// Once a response is received from a remote capability further transmission should be cancelled
+	ctx, cancelFn := context.WithCancel(parentCtx)
+	defer cancelFn()
+
+	if err := c.transmitRequestWithMessageID(ctx, req, deterministicMessageID); err != nil {
+		return nil, fmt.Errorf("failed to transmit request: %w", err)
+	}
+
+	select {
+	case <-responseReceived:
+
+		response, loaded := c.messageIDToResponse.LoadAndDelete(deterministicMessageID)
+		if !loaded {
+			return nil, fmt.Errorf("no response found for message ID %s", deterministicMessageID)
+		}
+
+		msg, ok := response.(*types.MessageBody)
+		if !ok {
+			return nil, fmt.Errorf("unexpected response type %T for message ID %s", response, deterministicMessageID)
+		}
+
+		if msg.Error != types.Error_OK {
+			return nil, fmt.Errorf("remote capability returned error: %s", msg.Error)
+		}
+
+		capabilityResponse, err := pb.UnmarshalCapabilityResponse(msg.Payload)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal capability response: %w", err)
+		}
+
+		// TODO handle the case where the capability returns a stream of responses
+		resultCh := make(chan commoncap.CapabilityResponse, 1)
+		resultCh <- capabilityResponse
+		close(resultCh)
+
+		return resultCh, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+}
+
+// transmitRequestWithMessageID transmits a capability request to remote capabilities according to the transmission configuration
+func (c *remoteTargetCaller) transmitRequestWithMessageID(ctx context.Context, req commoncap.CapabilityRequest, messageID [32]byte) error {
+	rawRequest, err := pb.MarshalCapabilityRequest(req)
+	if err != nil {
+		return fmt.Errorf("failed to marshal capability request: %w", err)
+	}
+
+	// TODO should the transmission config be passed into the constructor rather than pulled from the request?
+	tc, err := transmission.ExtractTransmissionConfig(req.Config)
+	if err != nil {
+		return fmt.Errorf("failed to extract transmission config from request config: %w", err)
+	}
+
+	message := &types.MessageBody{
+		CapabilityId:    c.remoteCapabilityInfo.ID,
+		CapabilityDonId: c.remoteCapabilityDonInfo.ID,
+		CallerDonId:     c.localDONInfo.ID,
+		Method:          types.MethodExecute,
+		Payload:         rawRequest,
+		MessageId:       messageID[:],
+	}
+
+	peerIDToDelay, err := transmission.GetPeerIDToTransmissionDelay(c.remoteCapabilityDonInfo.Members, c.localDONInfo.Config.SharedSecret, req.Metadata.WorkflowID, req.Metadata.WorkflowExecutionID, tc)
+	if err != nil {
+		return fmt.Errorf("failed to get peer ID to transmission delay: %w", err)
+	}
+
+	for peerID, delay := range peerIDToDelay {
```

Comment:
It's interesting that you put the strategy inside the shim. I initially though that it will exist outside of it but maybe this is better. Let me think about it more.